compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -865,46 +853,121 @@ struct DistributeContract final : OpDistributionPattern<vector::ContractionOp> {
+    partialReductionShape.reserve(lhsLayout.getRank());
     if (resVector) {
-      replaceOpWithDistributedValues(rewriter, contractOp, accReduced);
-    } else {
-      Value accReducedVal = rewriter.create<vector::ExtractOp>(
-          loc, accReduction, SmallVector<int64_t>{0});
-      replaceOpWithDistributedValues(rewriter, contractOp, accReducedVal);
+      ArrayRef<int64_t> resVectorShape = resVector.getType().getShape();
+      partialReductionShape.insert(partialReductionShape.end(),
+                                   resVectorShape.begin(),
+                                   resVectorShape.end());
+    }
+    ArrayRef<int64_t> lhsShape = contractOp.getLhs().getType().getShape();
+    // Note that partial reduction dimensions are at the end.
+    SmallVector<int64_t> partialReduceDims;
+    partialReduceDims.reserve(reducedDims.size());
+    for (int64_t rDim : reducedDims) {
+      partialReduceDims.push_back(partialReductionShape.size());
+      partialReductionShape.push_back(lhsShape[rDim]);
     }
+    VectorType unDistributedType = VectorType::get(
+        partialReductionShape, localContractValue.getType().getElementType());
+    Value undistrLocalReduced = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+        loc, unDistributedType, isoRankLocalReduced);
 
+    // Manually infer the layout of partial reduction
+    IREE::VectorExt::NestedLayoutAttr reductionLayout;
+    {
+      SmallVector<int64_t> subgroupTileLens =
+          llvm::to_vector(resLayout.getSubgroupTile());
+      SmallVector<int64_t> batchTileLens =
+          llvm::to_vector(resLayout.getBatchTile());
+      SmallVector<int64_t> outerTileLens =
+          llvm::to_vector(resLayout.getOuterTile());
+      SmallVector<int64_t> threadTileLens =
+          llvm::to_vector(resLayout.getThreadTile());
+      SmallVector<int64_t> elementTileLens =
+          llvm::to_vector(resLayout.getElementTile());
+      SmallVector<int64_t> subgroupStrides =
+          llvm::to_vector(resLayout.getSubgroupStrides());
+      SmallVector<int64_t> threadStrides =
+          llvm::to_vector(resLayout.getThreadStrides());
+      for (int64_t rDim : reducedDims) {
+        // thread-local reductions have already been carried out.
+        // What is remaining is reductions across threads and
+        // subgroups.
+        subgroupTileLens.push_back(lhsLayout.getSubgroupTile()[rDim]);
+        batchTileLens.push_back(1);
+        outerTileLens.push_back(1);
+        threadTileLens.push_back(lhsLayout.getThreadTile()[rDim]);
+        elementTileLens.push_back(1);
+        subgroupStrides.push_back(lhsLayout.getSubgroupStrides()[rDim]);
+        threadStrides.push_back(lhsLayout.getThreadStrides()[rDim]);
+      }
+      reductionLayout = IREE::VectorExt::NestedLayoutAttr::get(
+          contractOp.getContext(), subgroupTileLens, batchTileLens,
+          outerTileLens, threadTileLens, elementTileLens, subgroupStrides,
+          threadStrides);
```

Comment:
Can we add a "broadcast" function to the layout which takes a dimension and sizes for a single dimension? This would make it easier to write these transformations.